### PR TITLE
Bessel zero finding cleanup

### PR DESCRIPTION
**Pull request by chris-langfield**
The functions `num_besselj_zeros`, `besselj_zeros`, `check_besselj_zeros` are almost exact ports from the MATLAB code, but are lacking clear comments.

These in turn are modifications of this script by Jonas Lundgren:

https://www.mathworks.com/matlabcentral/mlc-downloads/downloads/submissions/26639/versions/5/previews/zerobess.m/index.html

When attempting to work out what exactly these functions are doing to estimate the first few zeros, I came across several tricks that I haven't been able to find a single definitive source for. e.g.,

https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/6cb389edf21e811224cc98f4fde0d63ed778b9a7/src/aspire/basis/basis_utils.py#L165-L174

Roughly, this computes initial guesses based on powers of `nu`, the real number order of the Bessel function. The closest literature I could find for this is in the following paper:

Elbert, Árpád. “Some recent results on the zeros of Bessel functions and orthogonal polynomials.” Journal of Computational and Applied Mathematics 133 (2001): 65-83.

(Section 1.3 starting around eqn 1.7)

The powers as well as the hardcoded values in our code also appear here. 

See also: https://www.boost.org/doc/libs/1_63_0/libs/math/doc/html/math_toolkit/bessel/bessel_root.html

It may be a lesser concern to be able to find the theoretical justification here, but it would probably be ideal if we could. Regardless, I'm going to use this PR to comment and clean up these functions as much as possible so that it is easier to understand in the future

### CODE REVIEW

**Comment by chris-langfield** [comment]
It would be good to have another pair of eyes on this line:

https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/6cb389edf21e811224cc98f4fde0d63ed778b9a7/src/aspire/basis/basis_utils.py#L27-L31

`np.spacing` returns the distance to the nearest machine-representable float. My understanding of this check is that `ddz`, the second order differences between the zeros, should be zero up to 16x machine precision.

**Comment by chris-langfield** [comment]
Created #597 for continuing conversation, as we can merge this PR with just the docstrings maybe

**Review by j-c-c — DISMISSED** [review]
Looks good. Much clearer!

File: src/aspire/basis/basis_utils.py [hunk 1]
```diff
@@ -16,23 +16,48 @@
 
 
 def check_besselj_zeros(nu, z):
+    """
+    Verify a sequence of estimated zeros of the Bessel function with order `nu`.
```
Comment (janden) [inline src/aspire/basis/basis_utils.py]: Verify in what sense?

File: src/aspire/basis/basis_utils.py [hunk 2]
```diff
@@ -16,23 +16,48 @@
...
     else:
         result = result and all(ddz > -16 * np.spacing(z[1:-1]))
 
     return result
 
 
 def besselj_newton(nu, z0, max_iter=10):
+    """
+    Uses the Newton-Raphson method to compute the zero(s) of the
+    Bessel function of the First Kind with order `nu` with initial
```
Comment (janden) [inline src/aspire/basis/basis_utils.py]: Skip “of the First Kind”. We don't deal with any other Bessel functions here.

Same below.

File: src/aspire/basis/basis_utils.py [hunk 3]
```diff
@@ -216,12 +250,25 @@ def besselj_zeros(nu, k):
...
 
 def num_besselj_zeros(ell, r):
+    """
+    Compute the zeros of the order `ell` Bessel function of
+    the First Kind, which are less than `r`.
+
+    :param ell: The real number order of the Bessel function.
+    :param r: The upper bound for zeros returned.
+    :return n, r0: The number of zeros and the zeros themselves
+    as a numpy array.
```
Comment (janden) [inline src/aspire/basis/basis_utils.py]: NumPy

File: src/aspire/basis/basis_utils.py [hunk 4]
```diff
@@ -157,6 +182,15 @@ def real_sph_harmonic(j, m, theta, phi):
 
 
 def besselj_zeros(nu, k):
+    """
+    Finds the first `k` zeros of the Bessel function of the First Kind
+    of order `nu`, i.e. J_nu.
+    Adapted from "zerobess.m" by Jonas Lundgren <splinefit@gmail.com>
+
+    :param nu: The real number order of the Bessel Function. (must be positive and <10e7)
```
Comment (janden) [inline src/aspire/basis/basis_utils.py]: Lowercase “Function”. Move parentheses into sentence. Replace `10e7` with `1e7`.

**Review by janden — CHANGES_REQUESTED** [review]

**Comment by janden** [comment]
> It would be good to have another pair of eyes on this line:
> 
> https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/6cb389edf21e811224cc98f4fde0d63ed778b9a7/src/aspire/basis/basis_utils.py#L27-L31
> 
> `np.spacing` returns the distance to the nearest machine-representable float. My understanding of this check is that `ddz`, the second order differences between the zeros, should be zero up to 16x machine precision.

I think it's rather checking that the second-order differences are zero or *negative*. If they were all zero, we'd have a three-term recurrence formula for the zeros and wouldn't need to find them this way (although maybe that's true for large zeros).

**Comment by janden** [comment]
> (although maybe that's true for large zeros)

Sorry I should say asymptotically. It might be a decent approximation.

File: src/aspire/basis/basis_utils.py (hunk 1 above)
Comment (chris-langfield) [reply src/aspire/basis/basis_utils.py]: Yeah it doesn't really directly verify in the sense of computing `jv(z)`. It checks that zeros are positive, real, spaced greater than 3 apart, and a few other properties. Do you think I should change this word, or add some kind of additional check like this?

**Comment by janden** [comment]
Right. Of course. From Zhao & Singer (2013):

image

So at large `q` the approximate spacing between zeros is π. In fact, we use this in https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/6cb389edf21e811224cc98f4fde0d63ed778b9a7/src/aspire/basis/basis_utils.py#L193 to initialize the search for zeros by using the last zero and adding multiples of π when the number of zeros is large enough.

File: src/aspire/basis/basis_utils.py (hunk 1 above)
Comment (janden) [reply src/aspire/basis/basis_utils.py]: Might just call it a sanity check, since, as you say, it doesn't actually verify that they're zeros.

**Comment by chris-langfield** [comment]
Added your comments to #597 and made the docstring changes to this PR. Thanks!

**Review by janden — APPROVED** [review]
LGTM!

**Pull request merged**